File: plugin/com.hello.uxml.tools.ui/src/com/hello/uxml/tools/framework/LineShape.java
package com.hello.uxml.tools.framework;

import com.hello.uxml.tools.framework.graphics.UISurface;

import java.util.EnumSet;

/**
 * Defines a line shape
 *
 * @author ferhat@
 */
public class LineShape extends Shape {

  /** Start x axis coordinate */
  public static PropertyDefinition xFromPropDef = PropertySystem.register("XFrom",
      Double.class, LineShape.class, new PropertyData(0.0, null, EnumSet.of(
          PropertyFlags.Resize)));

  /** Start y axis coordinate */
  public static PropertyDefinition yFromPropDef = PropertySystem.register("XFrom",
      Double.class, LineShape.class, new PropertyData(0.0, null, EnumSet.of(
          PropertyFlags.Resize)));

  /** End x axis coordinate */
  public static PropertyDefinition xToPropDef = PropertySystem.register("XTo",
      Double.class, LineShape.class, new PropertyData(0.0, null, EnumSet.of(
          PropertyFlags.Resize)));

  /** End y axis coordinate */
  public static PropertyDefinition yToPropDef = PropertySystem.register("YTo",
      Double.class, LineShape.class, new PropertyData(0.0, null, EnumSet.of(
          PropertyFlags.Resize)));

  /**
   * Sets or returns start x coordinate.
   */
  public double getXFrom() {
    return ((Double) getProperty(xFromPropDef)).doubleValue();
  }

  public void setXFrom(double value) {
    setProperty(xFromPropDef, value);
  }

  /**
   * Sets or returns start y coordinate.
   */
  public double getYFrom() {
    return ((Double) getProperty(yFromPropDef)).doubleValue();
  }

  public void setYFrom(double value) {
    setProperty(yFromPropDef, value);
  }

  /**
   * Sets or returns end x coordinate.
   */
  public double getXTo() {
    return ((Double) getProperty(xToPropDef)).doubleValue();
  }

  public void setXTo(double value) {
    setProperty(xToPropDef, value);
  }

  /**
   * Sets or returns end y coordinate.
   */
  public double getYTo() {
    return ((Double) getProperty(yToPropDef)).doubleValue();
  }

  public void setYTo(double value) {
    setProperty(yToPropDef, value);
  }

  /**
   * Returns size of line bounds.
   */
  @Override
  protected void onMeasure(double availableWidth, double availableHeight) {
    setMeasuredDimension(Math.abs(getXFrom() - getXTo()), Math.abs(getYFrom() - getYTo()));
  }

  @Override
  protected void onRedraw(UISurface surface) {
    surface.drawLine(getStroke(), getXFrom(), getYFrom(), getXTo(), getYTo());
  }
}

File: framework/lib/line_shape.dart
part of uxml;

/**
 * Define a line shape.
 *
 * @author ferhat@ (Ferhat Buyukkokten)
 */
class LineShape extends Shape {
  static ElementDef lineShapeElementDef;
  /** Start X coordinate property definition */
  static PropertyDefinition xFromProperty;
  /** Start Y coordinate property definition */
  static PropertyDefinition yFromProperty;
  /** End X coordinate property definition */
  static PropertyDefinition xToProperty;
  /** End Y coordinate property definition */
  static PropertyDefinition yToProperty;
  /**
   * Constructor.
   */
  LineShape() : super() {
  }

  /** Sets or returns start x coordinate */
  num get xFrom => getProperty(xFromProperty);

  set xFrom(num value) {
    setProperty(xFromProperty, value);
  }

  /** Sets or returns  start y coordinate */
  num get yFrom => getProperty(yFromProperty);

  set yFrom(num value) {
    setProperty(yFromProperty, value);
  }

  /** Sets or returns  end x coordinate */
  num get xTo => getProperty(xToProperty);

  set xTo(num value) {
    setProperty(xToProperty, value);
  }

  /** Sets or returns  end y coordinate */
  num get yTo => getProperty(yToProperty);

  set yTo(num value) {
    setProperty(yToProperty, value);
  }

  /**
   * Overrides UIElement.onMeasure to return size of path.
   */
  bool onMeasure(num availableWidth, num availableHeight) {
    setMeasuredDimension((xFrom - xTo).abs(), (yFrom - yTo).abs());
    return false;
  }

  /** Overrides UIElement.onRedraw. */
  void onRedraw(UISurface surface) {
    surface.clear();
    surface.drawLine(fill, stroke, xFrom, yFrom, xTo, yTo);
  }

  /** @see UxmlElement.getDefinition. */
  ElementDef getDefinition() => lineShapeElementDef;

  /** Registers component. */
  static void registerLineShape() {
    xFromProperty = ElementRegistry.registerProperty("XFrom",
        PropertyType.NUMBER, PropertyFlags.REDRAW | PropertyFlags.RESIZE, null,
        0.0);
    yFromProperty = ElementRegistry.registerProperty("YFrom",
        PropertyType.NUMBER, PropertyFlags.REDRAW | PropertyFlags.RESIZE, null,
        0.0);
    xToProperty = ElementRegistry.registerProperty("XTo",
        PropertyType.NUMBER, PropertyFlags.REDRAW | PropertyFlags.RESIZE, null,
        0.0);
    yToProperty = ElementRegistry.registerProperty("YTo",
        PropertyType.NUMBER, PropertyFlags.REDRAW | PropertyFlags.RESIZE, null,
        0.0);
    lineShapeElementDef = ElementRegistry.register("LineShape",
        Shape.shapeElementDef, [xFromProperty, yFromProperty,
        xToProperty, yToProperty], null);
  }
}
